Major update of CONTRIBUTING.md (#681)
kernelkit/infix: .github/CONTRIBUTING.md

Contributing to Infix
=====================

Thank :heart: you for taking the time to read this!

We welcome any help in the form of bug reports, fixes, or patches to add
new features.  We *prefer* GitHub pull requests, but are open to other
forms of collaboration as well.  [Let's talk!][support] :handshake:

If you are unsure how to start implementing an idea or fix:

 - :bug: open an issue, there are human friendly templates for _bugs_
   and _feature requests_
 - :speech_balloon: use the [Q&A Forum][discuss]

> _Talking about code and problems first is often the best way to get
> started before submitting a pull request.  We have found it always
> saves time, yours and ours._

:sparkles: General Guidelines
-----------------------------

> When submitting a bug reports or patches to bugs, please state which

Reviewer: ```suggestion
When submitting bug reports or patches to bugs, please state which
```

version the change is made against, what it does, and, more importantly
*why* -- from your perspective, why is it a bug, why does the code need
changing in this way.  Start with why.

 - :bug: Bug reports need metadata like Infix version or commit hash
 - :adhesive_bandage: Bug fixes also need version, and (preferably) a
   corresponding issue number for the ChangeLog
 - :new: New features, you need to get approval of the YANG model first!  
   :speech_balloon: Please use the [Forum][discuss], e.g., category:
   *Ideas*, or open a :pray: feature request issue
 - :white_check_mark: New features also need new regression tests, this
   can be basic tests or more complex use-case tests comprising multiple
   subsystems, see [Testing Changes](#testing-changes), below

Please take care to ensure you follow the project coding style and the
commit message format.  If you follow these recommendations you help

> the maintainer(s) and make it easier for them to include your code.

Reviewer: ```suggestion
the maintainers and make it easier for them to include your code.
```

> code!  Legacy takes precedence, and remember, we read code a lot more
> than write it, so legibility is important.

The ChangeLog deserves a separate mention:

 - Releases are listed in reverse chronological order order, so the
   latest/next release is at the beginning of the file
 - Only *user-facing bugs and features* are detailed, so code refactor,
   new tests, etc. are not listed.
 - Add your changes/features to the Changes section
 - Add your Fix line in the Fixes section, in numeric order 
 - Changes and fixes without an issue number are listed after all
   numbered ones
 - YANG model changes are documented in their respective model, for
   standard models, e.g., for `ietf-interfaces.yang`, the corresponding
   `infix-interfaces.yang` detail augments/deviations as revisions.

A final note, lines of code are allowed to be longer than 72 characters
these days, unless you live by PEP-8 (see above).  There is no enforced
maximum, but the team usually keep it around 100 characters for both C
and Python.

:test_tube: Testing Changes
---------------------------

Please test your changes, no matter how trivial or obviously correct

> they may seem.  Nobody is infallible, making mistakes is only human.

Reviewer: ```suggestion
they may seem.  Nobody is infallible, making mistakes is only human.
It is also the best insurance policy for your feature, since it
ensures that your use-case will remain functional as the source base
evolves.
```

     - Typically a hyphen or asterisk is used for the bullet, preceded
       by a single space, with blank lines in between, but conventions
       vary here

	If you use an issue tracker, put references to them at the bottom,
	like this:

	Resolves: #123
	See also: #456, #789

    Signed-off-by: Jacky Linker <jacky.linker@example.com>

This is an example of how to [automatically close][[closing]] an issue
when the commit is merged to mainline.  Several keywords are available.

:twisted_rightwards_arrows: Pull Requests
-----------------------------------------

> _The git repository is the canonical location for information._

A pull request should preferably address a single issue or change.  This
may of course include multiple related changes, but what is important to
remember is the bandwidth of the maintainers.  A smaller well documented
PR is more likely to get attention quicker in some reviewer's busy day.

> Well documented here means that each commit message stand on its own,

Reviewer: ```suggestion
Well documented here means that each commit message stands on its own,
```
